### ISSUE TITLE
Prove BaGet's registration implementation

### ISSUE THREAD

**Issue by loic-sharma** [loic-sharma/BaGet#71]
BaGet's registration implementation is vastly different from nuget.org's. Both have pros and cons, but nuget.org cannot move to BaGet's registration implementation until BaGet is proven. This requires:

* [ ] Investigate BaGet's reliability
  * [ ] Investigate how to get even close to 16 9's of uptime
  * [ ] Investigate true geo-redundancy
  * [ ] Investigate Content Delivery Network caching for reliability
* [ ] Investigate BaGet's read performance
  * [ ] Scale to 100 requests per second for parity
  * [ ] Stretch goal: scale to 1,000 requests per second 

# A Comparison of BaGet vs nuget.org

## nuget.org

nuget.org generates static JSON files for each registration artifacts. These files are uploaded to Azure Blob Storage using geo-redundant storage and are served behind a Content Delivery Network.

* Pros
  * **Reliable**. Azure Blob Storage is very reliable (16 9's of uptime using Geographically Redundant Storage).
  * **High read performance**. The registration is just a bunch of static files that can be served directly. This can scale to infinity.
* Cons
  * **Hard to change**. Adding a new field to the registration necessitates a rebuild of the registration.
  * **Hard to verify**. Each JSON file must be verified independently. nuget.org has a job that crawls through artifacts to verify that they contain the same data as the database.
  * **Low write performance**. It is difficult to parallelize the job that updates the registration. This hampers write performance.
  * **Blocking**. Bugs in the job that updates the registration has cause the ingestion pipeline to halt.

**Note**: Although nuget.org has multiple copies of its V3 resources, each registration copy is not truly geo-redundant. This isn't listed as a con as this issue is fixable with the current implementation.

## BaGet

BaGet stores its package metadata in a database. Each request is served by a BaGet service that queries the database for the metadata and then serializes it into JSON.

* Pros
  * **Easy to change**. A new field field can be added to the registration by deploying a new version of BaGet.
  * **Write performance**. BaGet can index many packages in parallel. The database handles all concurrency issues.
 * **Verifiable**. The database is the source of truth and its package metadata has constraints.
* Cons
  * **Reliability**. 16 9's of uptime is incredibly difficult to achieve for a service. This can be mitigated with geo-redundancy. Uptime is nuget.org's number one concern, thus, this is the primary blocking for adopting BaGet's registration implementation.
  * **Read performance**. As each read requires a database query, read throughput is limited by the database. Caching can be used to mitigate this.

**Comment by WernerMairl** [comment]
i like that i like that challenge ;-)

A few questions and assumtions to confirm (or not)...

Azure Blob Storage + CDN (used by NuGet.org) sounds good. 
Assuming that the decision was driven by the [Blob Storage as Static website feature]( https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blob-static-website) and a CDN on top.

CDN means someone like Cloudflare years ago, 
in the meantime a [native Azure solution is available](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-https-custom-domain-cdn) is available to.

From my last projects and lessons learned, i can confirm that Azure Blob Storage is perfect for mass data BUT it likes big files (4.77TB maximum file size) but it does'nt like high number of small files
(Azure Blob Metadata handling is slow)
That's may be the main reason for some of the "Cons" listed above.

Why not usung Azure Table Storage, with a self written file delivery on top (basically like storage blob static website) and a CDN that "protects" this service ?

Azure Table Storage is optimized for "smaller" chunks (entities) but with a high number (looks like our scenario)

assuming that the Nuget.org" team has investigated thinks like that... can you tell us something about the results/findings ?

btw. can you point me to the most critical controller/methods (RegistrationIndexController).
we need a set of performance tests (specs) for that.....

regards
Werner

**Comment by WernerMairl** [comment]
small steps: 
As first i need a better understanding of registration.....
refactoring and testing is my personal way for deep investigations....

please validate if you can accept #202 . 
the work there was done during investigations....

**Comment by dimaromantest** [comment]
wat is postgres format Database__ConnectionString
